src/parser.js

```diff
@@ -0,0 +1,782 @@
+// interface.
+class Parser {
+  constructor({minim, source, generateSourceMap}) {
+    // Parser options
+    this.minim = minim;
+    this.source = source;
+    this.generateSourceMap = generateSourceMap;
+
+    // Loaded, dereferenced Swagger API
+    this.swagger = null;
+    // Refract parse result
+    this.result = null;
+    // Refract API category
+    this.api = null;
+    // State of the current parsing path
+    this.path = [];
+    // Current resource group, if any
+    this.group = null;
+  }
+
+  parse(done) {
+    const {
+      Copy, Category, Member: MemberElement, ParseResult, SourceMap,
+    } = this.minim.elements;
+
+    const parser = new SwaggerParser();
+    const parseResult = new ParseResult();
+    this.result = parseResult;
+
+    let loaded;
+    try {
+      loaded = _.isString(this.source) ? yaml.safeLoad(this.source) : this.source;
+    } catch (err) {
+      this.makeAnnotation(ANNOTATIONS.CANNOT_PARSE, null,
+        'Problem loading the input');
+
+      if (err.mark) {
+        parseResult.first().attributes.set('sourceMap', [
+          new SourceMap([[err.mark.position, 1]]),
+        ]);
+      }
+
+      return done(null, parseResult);
+    }
+
+    // Some sane defaults since these are sometimes left out completely
+    if (loaded.info === undefined) {
+      loaded.info = {};
+    }
+
+    if (loaded.paths === undefined) {
+      loaded.paths = {};
+    }
+
+    // Parse and validate the Swagger document!
+    parser.validate(loaded, (err) => {
+      const swagger = parser.api;
+      this.swagger = parser.api;
+
+      if (err) {
+        if (swagger === undefined) {
+          return done(err, parseResult);
+        }
+
+        // Non-fatal errors, so let us try and create annotations for them and
+        // continue with the parsing as best we can.
+        if (err.details) {
+          const queue = [err.details];
+          while (queue.length) {
+            for (const item of queue[0]) {
+              this.makeAnnotation(ANNOTATIONS.VALIDATION_ERROR, item.path,
+                item.message);
+
+              if (item.inner) {
+                // TODO: I am honestly not sure what the correct behavior is
+                // here. Some items will have within them a tree of other items,
+                // some of which might contain more info (but it's unclear).
+                // Do we treat them as their own error or do something else?
+                queue.push(item.inner);
+              }
+            }
+            queue.shift();
+          }
+        }
+      }
+
+      const api = new Category();
+      this.api = api;
+      parseResult.push(api);
+
+      // Root API Element
+      api.classes.push('api');
+
+      if (swagger.info) {
+        this.path.push('info');
+
+        if (swagger.info.title) {
+          api.meta.set('title', swagger.info.title);
+
+          if (this.generateSourceMap) {
+            this.makeSourceMap(api.meta.get('title'), this.path.concat(['title']));
```

Comment:
This makes me feel that this `Parse` class is doing more than simply parsing, as in it has multiple responsibilities. Passing state between functions with `this` confuses me when I look at code :)

In other words, you could have a function that you pass a `Parse` instance into along with these other things and still have the same benefit without coupling the additional functionality to the instance.

This is mostly my opinion/preference. What are your thoughts?